File: Tree/510-lowestCommonAncestor.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        if not root:
            return 

        while root:
            if p.val<root.val and q.val<root.val:
                root=root.left
            elif p.val>root.val and q.val>root.val:
                root=root.right
            else:
                break;
        return root
    def lowestCommonAncestor2(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        def f(root,p,q):
            if not root:
                return 
            if p.val<root.val and q.val<root.val:
                return f(root.left,p,q)
            elif p.val>root.val and q.val>root.val:
                return  f(root.right,p,q)
            else:
                return root
        return f(root,p,q)
    

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution2:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        parent = {}
        def dfs(root, par):
            nonlocal parent
            if not root:
                return
            dfs(root.left, root)
            dfs(root.right, root)            
            parent[root] = par
            
        dfs(root, None)
        seen = set()
        for k,v in parent.items():
            if k and v:
                print(k.val,v.val)
        while p:      
            seen.add(p)            
            p = parent[p]
        
        while q: 
            if q in seen:
                return q
            q = parent[q]
        return None
    
    
six=TreeNode(6)
seven=TreeNode(7)
four=TreeNode(4)
five=TreeNode(5)
three=TreeNode(3)
two=TreeNode(2)
one=TreeNode(1)
eight=TreeNode(8)

two.left=one
two.right=three
five.left=four
five.right=six
four.left=two
seven.left=five

seven.right=eight

# print(Solution().lowestCommonAncestor(seven, one, three).val)
print(Solution2().lowestCommonAncestor(seven, one, three).val)

File: Tree/510-lowestCommonAncestor.cpp
#include "../treeNode.h"
#include<set>
#include<map>
#include<iostream>

using std::cout;
using SET=std::set<const TreeNode*>;
using MAP=std::map<const TreeNode*,const TreeNode*>;

const TreeNode* f(const TreeNode* root,const TreeNode* p, const TreeNode* q)
{
    if(!root)
        return nullptr;
    if((p->m_val<root->m_val) && (q->m_val<root->m_val))
    {
        return f(root->left,p,q);
    }
    else if((p->m_val>root->m_val) && (q->m_val<root->m_val))
    {
        return f(root->right,p,q);
    }
    else
        return root;
}
const TreeNode* lowestCommonA(const TreeNode* root,const TreeNode* p, const TreeNode* q)
{
    if(!root)
        return nullptr;
    return f(root,p,q);
}

const TreeNode* lowestCommonAn(const TreeNode* root,const TreeNode* p, const TreeNode* q )  //iterator method
{
    if(!root)
        return nullptr;

    while(root)
    {
        if((p->m_val<root->m_val) && (q->m_val<root->m_val))
            root=root->left;
        else if ((p->m_val>root->m_val) && (q->m_val<root->m_val))
            root=root->right;
        else
            break;
    }
    return root;
}

void dfoo(const TreeNode* root, const TreeNode* par,MAP& m)
{
    if(!root)
        return;
    dfoo(root->left,root,m);
    dfoo(root->right,root,m);
    m[root]=par;
}
const TreeNode* lowestCommonAnMap(const TreeNode* root,const TreeNode* p, const TreeNode* q )  //map
{
    MAP m{};
    SET seen{};
    dfoo(root,nullptr,m);

    while(p)
    {
        seen.insert(p);
        p=m[p];
    }
    while(q)
    {
        if(seen.find(q)!=seen.end())
            return q;
        q=m[q];
    }
    return nullptr;
}


int main()
{
    TreeNode* seven=new TreeNode(7);
    TreeNode* eight=new TreeNode(8);
    TreeNode* five=new TreeNode(5);
    TreeNode* four=new TreeNode(4);
    TreeNode* two=new TreeNode(2);
    TreeNode* one=new TreeNode(1);
    TreeNode* three=new TreeNode(3);

    seven->left=five;
    five->left=four;
    four->left=two;
    two->left=one;
    two->right=three;
    seven->right=eight;

    cout<<lowestCommonA(seven,one,three)->m_val;
    cout<<lowestCommonAn(seven,one,three)->m_val;
    cout<<lowestCommonAnMap(seven,one,three)->m_val;
}
